posthog/temporal/workflows/s3_batch_export.py

```diff
@@ -0,0 +1,342 @@
+import datetime as dt
+import json
+from dataclasses import dataclass
+from string import Template
+
+from aiohttp import ClientSession
+from django.conf import settings
+from temporalio import activity, workflow
+from temporalio.common import RetryPolicy
+
+from posthog.settings.base_variables import DEBUG, TEST
+from posthog.temporal.workflows.base import (
+    CreateBatchExportRunInputs,
+    PostHogWorkflow,
+    UpdateBatchExportRunStatusInputs,
+    create_export_run,
+    update_export_run_status,
+)
+
+INSERT_INTO_S3_QUERY_TEMPLATE = Template(
+    """
+    INSERT INTO FUNCTION s3({path}, $auth {file_format})
+    $partition_clause
+    """
+)
+
+SELECT_QUERY_TEMPLATE = Template(
+    """
+    SELECT $fields
+    FROM $table_name
+    WHERE
+        timestamp >= toDateTime({data_interval_start}, 'UTC')
+        AND timestamp < toDateTime({data_interval_end}, 'UTC')
+        AND team_id = {team_id}
+    """
+)
+
+TABLE_PARTITION_KEYS = {
+    "events": {
+        "hour": "toStartOfHour(timestamp)",
+        "day": "toStartOfDay(timestamp)",
+        "week": "toStartOfWeek(timestamp)",
+        "month": "toStartOfMonth(timestamp)",
+    }
+}
+
+
+@dataclass
+class S3InsertInputs:
+    """Inputs for ClickHouse INSERT INTO S3 function."""
+
+    bucket_name: str
+    region: str
+    key_template: str
+    team_id: int
+    data_interval_start: str
+    data_interval_end: str
+    file_format: str = "CSV"
+    table_name: str = "events"
+    partition_key: str | None = None
+    aws_access_key_id: str | None = None
+    aws_secret_access_key: str | None = None
+
+
+def build_s3_url(bucket: str, region: str, key_template: str, **template_vars):
+    """Form a S3 URL given input parameters.
+
+    ClickHouse requires an S3 URL with http scheme.
+    """
+    if not template_vars:
+        key = key_template
+    else:
+        key = key_template.format(**template_vars)
+
+    if TEST or DEBUG:
```

Comment:
Can we use `OBJECT_STORAGE_ENDPOINT` here instead?

Reply:
Not quite: `settings.OBJECT_STORAGE_ENDPOINT` is set to `http://localhost:19000` which is accessible via the host. However, the ClickHouse container does not share the network from the host, so it uses the service name. I made a comment explaining this right below.

Reply:
ah yup yup 💩